File: go/src/72_Edit_Distance.go
package src

func minDistance(word1 string, word2 string) int {
	memo := make([][]int, len(word1))
	for i := range len(word1) {
		memo[i] = make([]int, len(word2))
	}
	return dfs(word1, word2, 0, 0, memo)
}

func dfs(word1 string, word2 string, i int, j int, memo [][]int) int {
	if i == len(word1) {
		return len(word2) - j
	} else if j == len(word2) {
		return len(word1) - i
	} else if memo[i][j] != 0 {
		return memo[i][j]
	}
	if word1[i] == word2[j] {
		return dfs(word1, word2, i+1, j+1, memo)
	} else {
		del := dfs(word1, word2, i+1, j, memo)
		replace := dfs(word1, word2, i+1, j+1, memo)
		insert := dfs(word1, word2, i, j+1, memo)
		memo[i][j] = 1 + min(del, replace, insert)
		return memo[i][j]
	}
}

File: python/src/72_Edit_Distance.py
def minDistance(self, word1, word2):
    """
    :type word1: str
    :type word2: str
    :rtype: int
    """
    matrix = [[0 for _ in range(len(word2) + 1)] for _ in range(len(word1) + 1)]
    for i, char1 in enumerate(" " + word1):
        for j, char2 in enumerate(" " + word2):
            if i == 0 and j == 0:
                continue
            elif i == 0 or j == 0:
                matrix[i][j] = max(i, j)
            elif char1 == char2:
                matrix[i][j] = matrix[i - 1][j - 1]
            else:
                matrix[i][j] = 1 + min(matrix[i - 1][j - 1], matrix[i][j - 1], matrix[i - 1][j])
    return matrix[-1][-1]